<h1>ML Assignment 2 - 41117 (P1)</h1>

<b>Classify the email using binary classification method. Email Spam detection has two states:</b>
- Normal State Not Spam
- Abnormal State Spam<br>
Use KNN and SVM for classification and analyse their performance

Import Dependencies

In [59]:
import pandas as pd
import numpy as np

# Import preprocessing tools
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Import models
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC

# Import evaluation parameters
from sklearn.metrics import confusion_matrix, accuracy_score

Read dataset and get general statistics

In [60]:
df = pd.read_csv("emails.csv")
df.head()

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5172 entries, 0 to 5171
Columns: 3002 entries, Email No. to Prediction
dtypes: int64(3001), object(1)
memory usage: 118.5+ MB


Check for Null values

In [62]:
df.isnull().sum()

Email No.     0
the           0
to            0
ect           0
and           0
             ..
military      0
allowing      0
ff            0
dry           0
Prediction    0
Length: 3002, dtype: int64

Creating feature and target variables

In [63]:
X = df.iloc[:, 1:-1].values # Word frequencies 
Y = df.iloc[:, -1].values # Prediction column (0,1)

print(X, Y)

[[ 0  0  1 ...  0  0  0]
 [ 8 13 24 ...  0  1  0]
 [ 0  0  1 ...  0  0  0]
 ...
 [ 0  0  1 ...  0  0  0]
 [ 2  7  1 ...  0  1  0]
 [22 24  5 ...  0  0  0]] [0 0 0 ... 1 1 0]


Splitting into training and testing datasets

In [64]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)
print("X_train:", len(X_train))
print("Y_train:", len(Y_train))
print("X_test:", len(X_test))
print("Y_test:", len(Y_test))

X_train: 4137
Y_train: 4137
X_test: 1035
Y_test: 1035


Scaling Features

In [65]:
print(X_train[:5])

s = StandardScaler()
X_train = s.fit_transform(X_train)
X_test = s.transform(X_test)

print()
print(X_train[:5])

[[ 0  0  2 ...  0  1  0]
 [14 22  5 ...  0  4  0]
 [ 2  0  3 ...  0  2  0]
 [ 0  0  1 ...  0  0  0]
 [ 1  0  1 ...  0  0  0]]

[[-0.5645139  -0.64670049 -0.22485481 ... -0.06032421  0.01783856
  -0.07215751]
 [ 0.62527927  1.61622908 -0.01529762 ... -0.06032421  1.04281228
  -0.07215751]
 [-0.39454345 -0.64670049 -0.15500242 ... -0.06032421  0.35949647
  -0.07215751]
 [-0.5645139  -0.64670049 -0.29470721 ... -0.06032421 -0.32381935
  -0.07215751]
 [-0.47952868 -0.64670049 -0.29470721 ... -0.06032421 -0.32381935
  -0.07215751]]


<b>KNN MODEL</b><br>
- Counts 'k' neighbours of data points
- Data point is categorized into the category that contains most of the nearest neighbours

In [66]:
knn = KNeighborsClassifier()
knn.fit(X_train, Y_train)

KNeighborsClassifier()

In [67]:
Y_pred = knn.predict(X_test)
print("Y_test:", Y_test)
print("Y_pred:", Y_pred)

Y_test: [0 0 0 ... 1 0 0]
Y_pred: [1 0 0 ... 1 0 0]


In [74]:
cm1 = confusion_matrix(Y_test, Y_pred)
print("KNN Confusion Matrix:")
print(cm1)

acc1 = accuracy_score(Y_test, Y_pred)
print("\nKNN Accuracy:", acc1)

KNN Confusion Matrix:
[[586 164]
 [ 19 266]]

KNN Accuracy: 0.8231884057971014


<b>SVM MODEL</b><br>
- Tries to create a decision boundary / hyperplane for classification
- Can be linear or non-linear

In [75]:
sv = SVC()
sv.fit(X_train, Y_train)

SVC()

In [76]:
Y_pred = sv.predict(X_test)
print("Y_test:", Y_test)
print("Y_pred:", Y_pred)

Y_test: [0 0 0 ... 1 0 0]
Y_pred: [0 0 0 ... 1 0 0]


In [79]:
cm2 = confusion_matrix(Y_test, Y_pred)
print("SVM Confusion Matrix:")
print(cm2)

acc2 = accuracy_score(Y_test, Y_pred)
print("\n SVM Accuracy:", acc2)

SVM Confusion Matrix:
[[747   3]
 [ 64 221]]

 SVM Accuracy: 0.9352657004830918


<b>Accuracy:</b><br>
- <b>KNN: </b>0.823
- <b>SVM: </b>0.935
<br><br>
Hence SVM provided a better accuracy but took longer to compute.